src/Transform/ONNX/Recompose.cpp

```diff
@@ -137,97 +172,135 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
-      return reportFailure("epsilon is expected to be scalar");
+      return reportFailure("RMS epsilon is expected to be scalar");
     ONNXConstantOp epsilonOp =
         dyn_cast<ONNXConstantOp>(epsilon.getDefiningOp());
     if (!epsilonOp)
-      return reportFailure("epsilon needs to be a constant");
+      return reportFailure("RMS epsilon needs to be a constant");
     epsilonAttr = epsilonOp.getValueFloatAttr();
     // Check axes.
-    if (!hasShapeAndRank(x))
-      return reportFailure("need rank and shape for input x");
-    int64_t xRank = x.getType().cast<ShapedType>().getRank();
-    int64_t meanAxis, varAxis;
-    if (!suitableAxis(mReduceOp, xRank, meanAxis))
-      return reportFailure("unsuitable mean reduce axes");
-    if (!suitableAxis(vReduceOp, xRank, varAxis))
-      return reportFailure("unsuitable var reduce axes");
-    if (meanAxis != varAxis)
-      return reportFailure("mean and var axes must be the same");
-    // Axis is fine
-    axis = meanAxis;
-    LLVM_DEBUG(llvm::dbgs() << "LayerNorm from mult, axis : " << axis << "\n");
+    if (!hasShapeAndRank(dd))
+      return reportFailure("RMS need rank and shape for input dd");
+    int64_t ddRank = dd.getType().cast<ShapedType>().getRank();
+    int64_t varAxis;
+    if (!suitableAxis(vReduceOp, ddRank, varAxis))
+      return reportFailure("RMS unsuitable var reduce axes");
+
+    // 3: All the conditions are now correct for having an RMS pattern.
+    // Now check if we can extend the pattern to a full LM pattern.
+
+    bool hasFullPattern = true;
+    // %d = "onnx.Sub"(%X, %mean)
+    // %dd = "onnx.Mul"(%d, %d)
+    if (!operandOfOpDefinedBy<ONNXSubOp>(dSubOp, ddMulOp, d1, d2, 0))
+      hasFullPattern = reportFailure("LN missing D, sub op");
+    if (hasFullPattern &&
+        !operandOfOpDefinedBy<ONNXSubOp>(dSubOp, ddMulOp, d1, d2, 1))
+      hasFullPattern = reportFailure("LN missing D, sub op");
+    // %mean = "onnx.ReduceMeanV13"(%x)
+    // %d = "onnx.Sub"(%X, %mean)
+    if (hasFullPattern && !operandOfOpDefinedBy<ONNXReduceMeanV13Op>(
+                              mReduceOp, dSubOp, x1, mean, 1))
+      hasFullPattern = reportFailure("LN missing mean, reduce mean op");
+
+    // 4: We have the ops for a traditional LM pattern, now check a few more
+    // things.
+
+    // d = sub has 3 uses from 2 distinct ops, ignore test for now.
+
+    if (hasFullPattern) {
+      // Verify that the mReduceOp uses x as well.
+      auto lnOp = cast<ONNXReduceMeanV13Op>(mReduceOp);
+      Value x2 = lnOp.getData();
+      if (x1 != x2)
+        hasFullPattern = reportFailure(
+            "LN input x to mean/ReduceMean and sub are different");
+    }
+    // Check the number of uses, for now only the 1 uses.
+    if (hasFullPattern && !mReduceOp->hasOneUse())
+      hasFullPattern = reportFailure("LN mean reduce has too many uses");
+    if (hasFullPattern) {
+      if (!hasShapeAndRank(x1))
+        return reportFailure("LN need rank and shape for input x");
```

Comment:
Cool, this is quite clever. It causes another try when `x1` has shape to get the full pattern instead of greedily conclude that it is RMS pattern.